test-loop-tests/src/tests/catching_up.rs

```diff
@@ -0,0 +1,515 @@
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+
+use itertools::Itertools as _;
+use near_async::messaging::{CanSend as _, Handler as _};
+use near_async::time::Duration;
+use near_chain_configs::test_genesis::TestEpochConfigBuilder;
+use near_client::{ProcessTxRequest, Query};
+use near_network::types::NetworkRequests;
+use near_o11y::testonly::init_test_logger;
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::sharding::ChunkHash;
+use near_primitives::test_utils::create_user_test_signer;
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::types::{
+    AccountId, BlockHeight, BlockHeightDelta, BlockReference, NumSeats, ShardId,
+};
+use near_primitives::views::{QueryRequest, QueryResponseKind};
+
+use crate::setup::builder::TestLoopBuilder;
+use crate::utils::ONE_NEAR;
+use crate::utils::rotating_validators_runner::RotatingValidatorsRunner;
+use crate::utils::transactions::get_anchor_hash;
+
+/// Verifies that fetching of random parts works properly by issuing transactions during the
+/// third epoch, and then making sure that the balances are correct for the next three epochs.
+/// If random one parts fetched during the epoch preceding the epoch a block producer is
+/// assigned to were to have incorrect receipts, the balances in the fourth epoch would have
+/// been incorrect due to wrong receipts applied during the third epoch.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync() {
+    test_catchup_random_single_part_sync_common(false, false, 22)
```

Comment:
mini nit and only if you feel like it:
It's hard to say what does bools mean and it's easy to accidentally switch them around when authoring new tests. You can consider either of the following improvements:
1) Create enums for each with descriptive names e.g. ```enum SkipConfig{ NoSkip, Skip(BlockHeight) }```
2) Create a single struct consisting of all the relevant arguments as fields ```struct Catching { skip: bool, non_zero: bool, ...}```

Option 1) is a bit more ergonomic to use but 2) has a name for the third argument which is nice.